Library/Homebrew/cmd/bottle.rb

```diff
@@ -300,72 +301,86 @@ def bottle_formula(f)
       end
     end
   end
 
-  module BottleMerger
-    def bottle(&block)
-      instance_eval(&block)
-    end
-  end
-
   def merge
+    require "formulary"
+    Formulary.class_eval { define_method(:load_formula) { raise "Loaded formula!" } }
+
     write = ARGV.include? "--write"
-    keep_old = ARGV.include? "--keep-old"
-
-    merge_hash = {}
-    ARGV.named.each do |argument|
-      bottle_block = IO.read(argument)
-      formula_name = bottle_block.lines.first.sub(/^# /, "").chomp
-      merge_hash[formula_name] ||= []
-      merge_hash[formula_name] << bottle_block
+
+    bottles_hash = ARGV.named.reduce({}) do |hash, json_file|
+      hash.merge! Utils::JSON.load(IO.read(json_file))
     end
 
-    merge_hash.each do |formula_name, bottle_blocks|
+    bottles_hash.each do |formula_name, bottle_hash|
       ohai formula_name
-      f = Formulary.factory(formula_name)
 
-      if f.bottle_disabled?
-        ofail "Formula #{f.full_name} has disabled bottle"
-        puts f.bottle_disable_reason
+      if bottle_hash["bottle"]["disabled?"]
+        ofail "Formula #{formula_name} has disabled bottle"
+        puts bottle_hash[:bottle][:disable_reason]
         next
       end
 
-      bottle = if keep_old
-        f.bottle_specification.dup
-      else
-        BottleSpecification.new
+      bottle = BottleSpecification.new
+      bottle.root_url bottle_hash["bottle"]["root_url"]
+      cellar = bottle_hash["bottle"]["cellar"]
+      if cellar == "any" || cellar == "any_skip_relocation"
+        cellar = cellar.to_sym
       end
-      bottle.extend(BottleMerger)
-      bottle_blocks.each { |block| bottle.instance_eval(block) }
-
-      old_spec = f.bottle_specification
-      if keep_old && !old_spec.checksums.empty?
-         bad_fields = [:root_url, :prefix, :cellar, :revision].select do |field|
-           old_spec.send(field) != bottle.send(field)
-         end
-         bad_fields.delete(:cellar) if old_spec.cellar == :any && bottle.cellar == :any_skip_relocation
-         if bad_fields.any?
-           ofail "--keep-old is passed but there are changes in: #{bad_fields.join ", "}"
-           next
-         end
+      bottle.cellar cellar
+      bottle.prefix bottle_hash["bottle"]["prefix"]
+      bottle.revision bottle_hash["bottle"]["revision"]
+      bottle_hash["bottle"]["tags"].each do |tag, tag_hash|
+        bottle.sha256 tag_hash["sha256"] => tag.to_sym
       end
 
       output = bottle_output bottle
       puts output
 
       if write
+        path = Pathname.new("#{HOMEBREW_REPOSITORY/bottle_hash["formula"]["path"]}")
         update_or_add = nil
 
-        Utils::Inreplace.inreplace(f.path) do |s|
+        Utils::Inreplace.inreplace(path) do |s|
           if s.include? "bottle do"
             update_or_add = "update"
-            string = s.sub!(/  bottle do.+?end\n/m, output)
+            if ARGV.include? "--keep-old"
+              # TODO: read existing bottle block and append new output line to top.
```

Comment:
I prefer the bottle JSON to only contain the generated bottles. Besides, I think it's more secure to check this on both sides.